tutorial/stream-client.md

```diff
@@ -0,0 +1,276 @@
+
+```
+flask==3.0.3
+requests==2.32.3
+pyopenssl==24.2.1
+```
+
+Create a `service.py` file inside `stream-client` with the following code:
+
+```python
+import os
+import os.path
+import secrets
+
+from flask import Flask, jsonify, render_template, request
+from requests import Session
+from functools import wraps
+from typing import *
+
+
+class GatewayAPI:
+    def __init__(self, base_url, session, token, verify):
+        self.base_url = base_url
+        self.session = session
+        self.token = token
+        self.verify = verify
+
+    def make_request(
+        self,
+        method,
+        path,
+        headers,
+        data,
+    ):
+        uri = f"{self.base_url}{path}"
+
+        headers["Authorization"] = f"macaroon root={self.token}"
+        headers["Content-Type"] = "application/json"
+
+        return self.session.request(
+            method, uri, json=data, headers=headers, verify=self.verify
+        )
+
+    def join_session(self, session_id, data):
+        uri = f"/1.0/sessions/{session_id}/join"
+        return self.make_request(
+            "POST",
+            uri,
+            data=data,
+            headers={"Content-Type": "application/json"},
+        ).json()
+
+app = Flask(__name__)
+
+gateway_api_url = os.getenv("GATEWAY_URL")
+gateway_api_token = os.getenv("GATEWAY_API_TOKEN")
+gateway_server_cert = os.getenv("GATEWAY_SERVER_CERT") or False
+gateway_enabled = bool(gateway_api_url and gateway_api_token)
+
+auth_user = os.getenv("AUTH_USER") or "anbox"
+auth_password = os.getenv("AUTH_PASSOWRD") or secrets.token_urlsafe(16)
+
+print(f"You can login with username {auth_user} and password {auth_password}")
+
+gateway = GatewayAPI(gateway_api_url, Session(), gateway_api_token, gateway_server_cert)
+
+
+def render_error_response(msg, code):
+    response = {"error_msg": msg}
+    return jsonify(response), code
+
+
+def check_auth(username, password):
+    return username == auth_user and password == auth_password
+
+
+def login_required(f):
+    @wraps(f)
+    def wrapped_view(**kwargs):
+        auth = request.authorization
+        if not (auth and check_auth(auth.username, auth.password)):
+            return (
+                "Unauthorized",
+                401,
+                {"WWW-Authenticate": 'Basic realm="Login Required"'},
+            )
+        return f(**kwargs)
+
+    return wrapped_view
+
+
+@app.route("/1.0/sessions/<session_id>/join", methods=["POST"])
+@login_required
+def api_1_0_sessions_join_post(session_id):
+    if not gateway_enabled:
+        return render_error_response("no gateway connected", 503)
+
+    request_data = {"disconnect_clients": True}
+    response = gateway.join_session(session_id, request_data)
+
+    if "status_code" not in response or response["status_code"] != 200:
+        return render_error_response("failed to join session", 500)
```

Comment:
Maybe we want to include the error message in the response as well?
https://documentation.ubuntu.com/anbox-cloud/en/latest/reference/api-reference/gateway-api/#/session/handle-join-session

Reply:
Agreed, that should be added to the API documentation in swagger.

Reply:
I would like to do some more refinements on the tutorial once the first version is out where we can include such things. How does that sound @adglkh ?

Reply:
Sound good to me!